### DISCUSSION TITLE
How to use resources for non-streaming, static data

### DISCUSSION THREAD

**Discussion by ThomasFreedman** [markqvist/Reticulum#321] [Help & Questions] [answered]
***Mod edit: This great question is a good example of how not to use requests. The code in this post is wrong due to a misunderstanding of the API. Take a look at the answers for an explanation of why this is, and what the right thing to do is.***

***Additionally, resources are always used for non-streaming data, another incorrect assumption in the question. For streaming data, the Buffer and Channel classes can be used, but a Resource always transfers a bounded set of data.***

---

I wrote elsewhere:

> I don't need resources to stream, but rather just overcome the packet size limit. Just not sure about the callbacks & payload delivery. Seems like I need to handle packet payloads differently than resource payloads, but not sure. Trying to follow the file transfer example but that's more about streaming file data.

Reviewing the Understanding Reticulum, you wrote:
[Resources can] ...automatically handle sequencing, compression, coordination and checksumming.

It doesn't say assembling the parts, or if the handling pertains to individual parts or the total amount of data.

I tried to simply add my payload as a python dictionary to a resource and send that:
```
        msgResource = RNS.Resource(payload, link=pub["link"],
                                   timeout=self.MaxRNSwait,
                                   callback=msgAck)

        path = f"{self.AppName}/{msgType}"
        pub["link"].request(path, data=msgResource, 
                                       response_callback=responseRelay, 
                                       failed_callback=failRelay) 
```
but I get a: `TypeError: Invalid data instance type passed to resource initialisation`. So then I tried to use pickle.dumps and umsgpack.packb to convert the payload to a bytes obj for msgResource:
```
        msgResource = RNS.Resource(umsgpack.packb(payload), ...
        pub["link"].request(path, data=msgResource, ...
```
which also failed with a similar error. You didn't actually say so, but it seems the Resource object is intended for streaming bytes not python structures.

Is it possible to send a python object such as a dictionary or even a list that is bigger than a single packet will allow using the Resource object? I love that it works for smaller items that fit into a packet. Sure would be nice to get around that size limit with the same ease of use.

**Comment by markqvist** [comment]
Are you using the a `Resource` as part of a `Request`? That's not gonna work. You don't initialise and pass a `Resource` instance to the `request` call. You just pass the data/dict/structure/array/whatever.

**Reply by ThomasFreedman** [reply]
Of course not. I don't know how you got that impression from the snippets of code above. The first arg to the request call is "path" (`path = f"{self.AppName}/{msgType}"`), the second data, which is the resource. msgType is just a short string for the type of request being done. 

In terms of the total lines of code difference between having a generic request handler callback and my own dispatcher, vs using the path method you built into RNS of registering different handlers, it's almost identical. I changed my code to conform to your intended design pattern anyway, which saves a teeny tiny bit of data space. Why use any of the data space for the message type when you built in the path method of request routing?

What you said in your second answer is fantastic (that requests over a link aren't limited to the packet size). Not sure why I thought it was, but that's the impression I had from looking at RNS so far. I don't need Resources in that case. But just to be clear, are Resources *only* used for files or a streaming flow of bytes, or can they be used to transfer a python object that whose size requires multiple packets? (I recognize from your second answer there's no need for using Resources in that case, but is it possible?)

I do see a a limit imposed for Channels based on packet size though, which also require Links. That restriction is clearly called out in the Channel API description.

**Reply by markqvist** [reply, marked as answer]
A typo in the above post obscured what I was saying. Edited now.

Yes, you are using a `Resource` as part of a `Request`:

```python
msgResource = RNS.Resource(payload, link=pub["link"], timeout=self.MaxRNSwait, callback=msgAck)

pub["link"].request(path, data=msgResource, response_callback=responseRelay, failed_callback=failRelay) 
```

That's completely wrong. Nowhere in the docs is there any hint that you should do something like that ;) You **don't** use the `Resource` class when dealing with `Requests`. Reticulum will handle that internally. As I said before, you just pass whatever data you want to the `data` argument of the `request()` call. I repeat **do not** instantiate a `Resource` here.

Reticulum tries to keep things easy and simple, not completely obtuse ;)

What you do is just something like this:

```python
whatever_data = ... # Whatever you need to send with the request
link.request(path, data=whatever_data, response_callback=got_response, failed_callback=request_failed) 
```

You can pass a dict, an array, an integer, a binary blob, an open file handle, almost anything. But not a `Resource` instance.

**Reply by ThomasFreedman** [reply]
My apologies, I'm not communicating well.  I totally get how to use Request now, including the path. 

I did NOT realize the Resource object wasn't to be used *as a 'resource'* in a request. It is it's own thing. 

So am I right in presuming a Resource is only suited for byte streams and not static binary blobs in memory? That makes sense, now that I'm aware there is no size limitation on sending objects (binary or otherwise, binary if umsgpack is used). There is no need to use it for that when a request will suffice.

**Reply by markqvist** [reply]
> My apologies, I'm not communicating well. I totally get how to use Request now, including the path.

No worrries, questions like these shed some clarity on subjects that may confuse other people as well, it's very valuable.

> I did NOT realize the Resource object wasn't to be used _as a 'resource'_ in a request. It is it's own thing.

Yes, exactly. You can think of it as one step lower level API than requests and responses. A `Resource` gives you an abstract way to simply transfer any kind (and size) of data over a `Link`. For even higher level functionality, there is the `Request` API, which then wraps this is request/response mechanisms.

> So am I right in presuming a Resource is only suited for byte streams and not static binary blobs in memory? That makes sense, now that I'm aware there is no size limitation on sending objects (binary or otherwise, binary if umsgpack is used). There is no need to use it for that when a request will suffice.

No, fortunately that is not correct. A `Resource` is not limited to files, but can be used for anything, both stuff on disk and in memory. Examples:

```python
data_binary = b"some binary blob"
data_number = 137.035999084
data_array = [2, 3, 5, "hello", 99, "world"]
data_dict = { "name": "something", "id": 1283744, "sensor_states": [4, 83.3, 7, 12, -8733] }
data_file = open("some_file.dat", "rb")
```

All of the above variables (and many other types) can be passed to the `data` argument when creating a  `Resource` or carrying out a `Request`.
